### PR TITLE
Implement AD for ODE systems

### LINKED ISSUE #8: [Automatic differentiation] Extend AD workflow to Systems with external drivers
In #6 we developed an AD workflow for optimising the Lorenz System. 

I had a look at the leaf_thermal model and realised we haven't dealt with Drivers yet. The drivers are external data that force the system (eg temperature). Data at discrete time intervals are passed in to create a spline that can interpolate continuously. When we run advance_adaptive, the spline is used to look up the driver value at any time point. 

The spline interpolation is something we want to avoid backpropagating through. It's expensive to record on the AD tape and the values are fixed (deterministic) and aren't being optimised (ie the gradient with respect to the loss is 1). 

Because of how we've set up our AD workflow (fixed schedule), I think we'll be able to use quite a simple solution where we cache the interpolated values as part of `set_target` and modify `ode_rates` to use the cache instead.

```c++
void set_target {
        fit_times_ = times;
        targets_ = targets;
        obs_indices_ = obs_indices;
        
        // Cache drivers if system has them
        cache_drivers(fit_times_);
    }
```

```c++
ode_rates {
    T t_air;
    
    if (driver_cache) {
        // Use cached value
        t_air = *driver_cache[current_interval];
    } else {
        // Normal interpolation
        t_air = temperature_fn->evaluate(time);
    }
    
    // ... 
}
```

```c++
advance_target {
    // ... 
    
    for (size_t i = 1; i < fit_times_.size(); ++i) {
        // set cache for the current interval
        if (drivers) {
            system.set_driver_cache(i);
        } 
        
        solver.step_to(system, fit_times_[i]);
     //... 
}
```

The bits that are a little fiddly are 1) we have to evaluate the spline on the `fit_times_` schedule before we start recording on the tape (i.e. in `set_target`) and 2) manage the cache for each RK substep during the `step_to` call in `advance_target`. 

I haven't fully fleshed out how this works, but I think we should aim to make it fairly automatic, so that people writing Systems only need to include the `if` statement demonstrated in `ode_rates`.

### PR DESCRIPTION
**Pull request by aornugent**
Addresses #2 and includes a workplan for #3 (the Lorenz System).

I've included the XAD library, but note this has a license incompatibility with the odelia package. I think it's pretty trivial, so maybe we can move to AGPL-3 if we think XAD is the way to go, or find a different way to integrate them without vendoring in the source code.

Speaking of which - XAD does include a static file (Tape.cpp) which breaks a little bit of Daniel's goal to make odelia header only. There are header-only autodiff libraries (e.g. autodiff/autodiff), but I've gone with XAD so that we can use it's checkpointing capabilties in more complex models that require an adaptive solver. That's a problem for later.

To get around the linking issues, I've reorganised the package in https://github.com/traitecoevo/odelia/commit/78da4bbe7638febdf0304ded79a2a5cd63254a15 so that you can use `make` to recompile the Rcpp and `devtools::load_all()` to load any changes you make that don't modify Rcpp exported interfaces. Both are quite quick for a project this size.

The parabola example is neat but simple. For a fixed objective, you can get the loss (e.g. distance from a maxima) and its gradient at any point:

image

which is what you need to give to an optimiser. BFGS gets to the objective in ~ 2 steps.

image

The relevant part of the AD example is in inst/examples/parabola.hpp. I have outlined what extending this pattern to an ODE System might look like in docs/lorenz-ad-exercise.md

### CODE REVIEW

File: inst/include/odelia/ode_fit.hpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+FitResult compute_gradient_ic(
+    System& system,
+    const std::vector<double>& ic,
+    const std::vector<double>& times,
+    const std::vector<std::vector<double>>& targets,
+    const std::vector<size_t>& obs_indices,
+    OdeControl& control
+) {
+    using ad = xad::adj<double>;
+    using ADType = ad::active_type;
```
Comment (aornugent) [inline inst/include/odelia/ode_fit.hpp]: `odelia` reserves PascalCase for objects like Solver, System and OdeControl, but lowercase for types. Change this to `ad_type` or `active` (I don't mind)

File: inst/include/odelia/ode_fit.hpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    
+    // 1. Create fresh tape
+    ad::tape_type tape;
+    
+    // 2. Convert IC to AD types and register as inputs
+    std::vector<ADType> ic_ad(ic.begin(), ic.end());
+    for (auto& x : ic_ad) {
+        tape.registerInput(x);
+    }
```
Comment (aornugent) [inline inst/include/odelia/ode_fit.hpp]: In order to combine `ics` and `params` in one go, we need some if statements. 

Because we're using the exisitng Solver, we need to get a reference to it's system. One method that doesn't exist yet is `set_params`. We also need to reset the system after changing any parameters or ICs.

```c++
    if (params) {
        std::vector<ad_type> params_ad(params->begin(), params->end());
        for (auto& p : params_ad) {
            tape.registerInput(p);
            inputs.push_back(p);
        }
        system.set_params(params_ad);
    }
    
    if (ic) {
        std::vector<ad_type> ic_ad(ic->begin(), ic->end());
        for (auto& x : ic_ad) {
            tape.registerInput(x);
            inputs.push_back(x);
        }
        system.set_ode_state(ic_ad.begin(), solver.fit_times()[0]);
    }
    
    system.reset()
```

File: inst/include/odelia/ode_fit.hpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    tape.newRecording();
+    
+    // 3. Set initial state
+    system.set_ode_state(ic_ad.begin(), times[0]);
+    
+    // 4. Create solver (no history collection to avoid tape pollution)
+    Solver<System> solver(system, control);
+    solver.set_collect(false);
```
Comment (aornugent) [inline inst/include/odelia/ode_fit.hpp]: Don't need this if we're passing in the Solver.

File: inst/include/odelia/ode_fit.hpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        // Check if current time index is an observation point
+        while (target_idx < obs_indices.size() && obs_indices[target_idx] == i) {
+            auto state = solver.state();
+            for (size_t j = 0; j < targets[target_idx].size() && j < state.size(); ++j) {
+                ADType diff = state[j] - ADType(targets[target_idx][j]);
+                loss += diff * diff;
+            }
+            target_idx++;
+        }
+    }
```
Comment (aornugent) [inline inst/include/odelia/ode_fit.hpp]: I think we can tidy this up substantially by creating a new solver method called `advance_target` (or something). Because the Solver already has `times` and `obs_indices` and knows when we want to calculate the loss, we just call `advance_target` and return the state at observation times.

```c++
    // Forward pass - returns states only at observation times
    auto obs = solver.advance_target();
    
    // Compute loss
    ad_type loss = sum_of_squares(obs, solver.targets());
```

One thing we discussed with Daniel is users wanting more flexibility with their losses, so it's a good idea to generalise it to a functor. Here's one example

```c++
template<typename T>
T sum_of_squares(const std::vector<std::vector<T>>& obs, 
                 const std::vector<std::vector<double>>& targets) {
    T loss(0.0);
    for (size_t i = 0; i < obs.size(); ++i) {
        for (size_t j = 0; j < obs[i].size(); ++j) {
            T diff = obs[i][j] - targets[i][j]; 
            loss += diff * diff;
        }
    }
    return loss;
}
```

One thing you might note is that obs is active type and targets are double. XAD supports mixed arithmetic, so the result will still be active type.

File: inst/include/odelia/ode_interface.hpp [hunk 5]
```diff
@@ -7,6 +7,31 @@
...
+    typename std::enable_if<
+        !std::is_same<typename System::value_type, void>::value
+    >::type> {
+    using value_type = typename System::value_type;
+    using state_type = std::vector<value_type>;
+};
+
+// 
+// Legacy typedefs (kept for backward compatibility with existing code)
+//
```
Comment (aornugent) [inline inst/include/odelia/ode_interface.hpp]: I have overcomplicated this, sorry. An easier solution is to say that all `odelia` systems expose a typedef.

This already exists in `lorenz_system.hpp`, but if we make it a rule (which I think we can) then we don't need traits or SFINAE.

```c++
template<typename T = double>
class LorenzSystem {
public:
    using value_type = T;  
    ...
};
```

If we accept this rule, then we can remove this section and make the change in the existing typedefs

```c++
// Before
typedef std::vector<double> state_type;

// After
template<typename System>
using state_type = std::vector<typename System::value_type>;
```

Effectively, we're creating the Solver's `state_type` based on the templated System it's intialised with. 

This is a much smaller change. All that happens later is we modify functions from using a static `state_type` to a templated `state_type<T>`. 

The exception is for functions that are exposed to R. These will always be `state_type<double>`.

File: inst/include/odelia/ode_interface.hpp [hunk 6]
```diff
@@ -111,49 +136,48 @@ ode_time(const T& /* obj */) {
 }
 
 namespace internal {
-template <typename T>
+template <typename T, typename StateType>
 typename std::enable_if<needs_time<T>::value, void>::type
-set_ode_state(T& obj, const state_type& y, double time) {
+set_ode_state(T& obj, const StateType& y, double time) {
   obj.set_ode_state(y.begin(), time);
 }
```
Comment (aornugent) [inline inst/include/odelia/ode_interface.hpp]: This becomes 

```c++
template <typename T>
typename std::enable_if<needs_time<T>::value, void>::type
set_ode_state(T& obj, const state_type<T>& y, double time) {
  obj.set_ode_state(y.begin(), time);
}
```

Where the only change is `state_type` -> `state_type<T>`.

File: inst/include/odelia/ode_solver.hpp [hunk 7]
```diff
@@ -40,10 +40,13 @@ class Solver
   // collectors
   double time() const { return solver.get_time(); }
 
-  ode::state_type state() const { return solver.get_state(); }
+  typename system_traits<System>::state_type state() const { return solver.get_state(); }
```
Comment (aornugent) [inline inst/include/odelia/ode_solver.hpp]: I think now that we've updated the state_type in `ode_interface.hpp` we can do

```c++
ode::state_type<System> state()
```

File: inst/include/odelia/ode_step.hpp [hunk 8]
```diff
@@ -12,6 +12,10 @@ namespace ode {
 template <class System>
 class Step {
 public:
+  // Extract scalar type from System using traits
+  using value_type = typename system_traits<System>::value_type;
+  using state_type = typename system_traits<System>::state_type;
+
```
Comment (aornugent) [inline inst/include/odelia/ode_step.hpp]: This can be

```c++
    using value_type = typename System::value_type;
    using state_type = std::vector<value_type>;
```

File: inst/include/odelia/solver_wrapper.hpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
+#ifndef ODELIA_SOLVER_WRAPPER_HPP_
```
Comment (aornugent) [inline inst/include/odelia/solver_wrapper.hpp]: I think we can remove this file if we change the R interface

File: inst/include/examples/lorenz_system.hpp [hunk 10]
```diff
@@ -2,82 +2,104 @@
...
+  // ODE interface
+  size_t ode_size() const { return ode_dimension; }
 
   double ode_time() const { return time; }
 
-  ode::const_iterator set_ode_state(ode::const_iterator it, double time_) {
-    
+  // Iterator methods remain templated on iterator type
+  template <typename Iterator>
+  Iterator set_ode_state(Iterator it, double time_) {
```
Comment (aornugent) [inline inst/include/examples/lorenz_system.hpp]: In order to adopt the changes to `compute_gradient` that reuse an existing Solver, we need the means to change it's parameters. We can model the pattern of `set_ode_state`

```c++

template <typename Iterator>
Iterator set_params(Iterator it) {
    sigma = *it++;
    R = *it++;
    b = *it++;
    return it;
}
```

**Review by aornugent — COMMENTED** [review]
Hi @Pushkalgithub -

Awesome work. My suggested changes below are only possible because we've seen how the overall system can work. They include some of the ideas we discussed yesterday with Daniel and some backtracking of how I originally thought we should do it.

Take a look through and let me know if you have any questions. I haven't tested them all, so it might not work exactly as planned, but hopefully you'll see where I think we can get to!

Feel free to ask questions or suggest improvements!

File: src/lorenz_interface.cpp [hunk 11]
```diff
@@ -41,36 +54,53 @@ inline Rcpp::XPtr<SystemType> get_system(SEXP xp) {
...
+    Rcpp::XPtr<SolverBaseType> ptr(wrapper, true);
+    return ptr;
+  } else {
+    // Create regular double solver
+    auto* sys_copy = new SystemType(*sys);
+    auto* wrapper = new ode::SolverWrapper<SystemType>(sys_copy, *ctrl);
+    
+    Rcpp::XPtr<SolverBaseType> ptr(wrapper, true);
+    return ptr;
+  }
```
[the remark this answers is not in the document]
Comment (dfalster) [reply src/lorenz_interface.cpp]: Nice!

**Comment by aornugent** [comment]
I was a bit overeager and introduced a different bug. The Lorenz system had an empty `reset` function, which means the solver state wasn't going back to t0. I could call `solver$fit` once, but the second time it would tell me my system was at the wrong timepoint. 

To get things untangled, I found it easiest to introduce `y_init` and `t0` in the system constructor, then have `system.reset` reinitialise the state and time. The side-effect of this is a cleaner `compute_gradient` which now doesn't mind if we aren't differentiating initial state (`ic`) - we used to have `compute_gradient` manage reinitialisiation, whereas now it's managed inside the system.

This changes the workflow slightly - before we'd create a system and call `lz$set_ode_state`. Now we call `lz$set_initial_state`.

I've added a test in `test-ad-workflow.R` to demonstrate what the interface looks like now.

@dfalster - I've resolved all but two comments. There's some code to convert `pars` and `record_step` output from active to double type, so they can be exposed in R. We don't record history with the active solver (because copying system state is an operation that gets recorded on the AD tape), and I don't think we need to expose `pars` as part of the AD workflow. I'm not sure if there's a chance someone accidentally calls it thought, so maybe there's a better guard we can put in place.

**Comment by dfalster** [comment]
Nice work @Pushkalgithub . Unfortunately I'm not getting to run. Are you able to successfully  run the code in `inst/examples/leaf_thermal/readme.qmd`?

On running 

```
Rcpp::sourceCpp("src/leaf_thermal_interface.cpp",
      verbose = TRUE, rebuild = TRUE
    )
```

I get errors
```
...
clang++ -arch arm64 -std=gnu++20 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/Library/Frameworks/R.framework/Resources/lib -L/opt/R/arm64/lib -o sourceCpp_5.so leaf_thermal_interface.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
Error in dyn.load("/private/var/folders/0x/nplts4jd5615dw_pr_np25540000gq/T/RtmpaOtbAS/sourceCpp-aarch64-apple-darwin20-1.1.1/sourcecpp_126c034eba94c/sourceCpp_5.so") : 
  unable to load shared object '/private/var/folders/0x/nplts4jd5615dw_pr_np25540000gq/T/RtmpaOtbAS/sourceCpp-aarch64-apple-darwin20-1.1.1/sourcecpp_126c034eba94c/sourceCpp_5.so':
  dlopen(/private/var/folders/0x/nplts4jd5615dw_pr_np25540000gq/T/RtmpaOtbAS/sourceCpp-aarch64-apple-darwin20-1.1.1/sourcecpp_126c034eba94c/sourceCpp_5.so, 0x0006): symbol not found in flat namespace '__ZN3xad4TapeIdLm1EE12active_tape_E'
  ```
  
  suggesting an issue with the `active_tape_`

**Comment by Pushkalgithub** [comment]
I was initially thinking about Andrew's suggestion to package external examples as mini-packages with `LinkingTo: odelia`, but I think the current approach with `odelia_load_dll()` + sourceCpp is actually better, it's simpler for users and avoids the complexity of managing multiple packages while still giving us clean external compilation.

**Comment by dfalster** [comment]
> One question: should we apply the same simplification pattern to the parabola example (use solver_interface.hpp), or leave it as-is for now?
> 
> @dfalster @aornugent

Good idea, but I think don't worry. I suggest that after we merge this branch, we delete parabola example. It will be in our history but has served it's purpose and is no longer needed

**Comment by aornugent** [comment]
Hia - I think the LeafThermal tests are working now. There was an issue with the `obs_indices` argument.

`Solver->set_target` takes a list of step times, some observations and a vector of indices that index into the step times. If we are running four steps, and are fitting to one observation at each step, then `obs_indices = c(1, 2, 3, 4)`.

I haven't fixed the Lorenz divergence in `test-ad-workflow.R` yet, sorry. They pass on 4266252ef5401ffcf44d0e732a62b128e9750e87 so there might be a change that's happened between now and then.

**Comment by dfalster** [comment]
Thanks for investigating and fixes @aornugent ! 

> They pass on https://github.com/traitecoevo/odelia/commit/4266252ef5401ffcf44d0e732a62b128e9750e87 so there might be a change that's happened between now and then.

Interesting! However, I went back to that commit just now and couldn't get it to compile.

**Comment by aornugent** [comment]
Great, thanks. 

Happy to leave the interface as is for now. Tabling for future reference that we could probably simplify some of the code for the active solver - rather than having every function handle both active and passive solvers, we could just scope some things to active (e.g. `fit`) and leave everything else (e.g. `collect`, `history` etc) to the passive solver.

I also wondered about this behaviour: https://github.com/traitecoevo/odelia/pull/6#issuecomment-4088071113

> I noticed recently that the Solver interface creates a new copy of a system when instantiating the solver, rather than using the pointer:
> 
> ```c++
> SEXP Solver_new(SEXP system_xp, SEXP control_xp, bool active = false) {
> ...
>     // Create regular solver
>     auto* sys_copy = new SystemType(*sys);
>     auto* solver = new ode::Solver<SystemType>(*sys_copy, *ctrl);
>     
>     return Rcpp::XPtr<ode::Solver<SystemType>>(solver, true);
> }
> ```

Rather than making copies, we could maybe use pointers so that if we change the System state / parameters, and re-run the solver, it creates a new trajectory. Similarly, when fitting, we could mutate the original System so that at the end of it's optimising, it contains the final values and is ready to generate the output trajectory.

Last thing would be to put in some guardrails for exploding losses. I noticed when fitting against Lorenz over `times = c(0, 10)` that the chaotic nature meant the optimiser fell over - small differences in parameters generated very large differences in trajectories. The test is constrained to the interval `[0, 1]` which is fine, but it'd be nicer for users if we handled bad losses or gradients more gracefully.

Like the interface, let's try these later as standalone issues with smaller PRs. I'm just getting them out of my head. 

---

Apart from the failing test on windows (something to do with accessing the installed DLLs?) I think this might be ready to land. We have demonstrated that we get valid gradients when optimising for initial conditions and for parameters in the LeafThermal system and demonstrated that we can optimise the parameters of the Lorenz system.

If you're happy, I'll do it this afternoon.

**Comment by dfalster** [comment]
Amazing work team! A great achievement. thanks for your hard work on this.
